### PR TITLE
Start a document for developers

### PR DESCRIPTION
**Pull request by blishko**
This document should collect more detailed information that could be useful for new developers of `hevm`.

### CODE REVIEW

File: doc/src/for-developers.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+Some of the interesting options are `-p <PATTERN>` to filter only some of the tests and `--quickcheck-tests <NUMBER>` to control how many tests quickcheck will generate for each property test.
+
+## On property-based testing
+
+There are a few ways to control how many tests Quickcheck will generate per property.
+By default, it generates 100 tests (satisfying the precondition).
+This can be controlled by `maxSuccess` argument passed to Quickcheck, or, in Tasty framework, using `localOption (QuickCheckTests <N>)`.
+Passing `--quickcheck-tests <N>` to the binary will change this value to `<N>`.
+This value can be dynamically adjustes for a test group or a specific test.
```
Comment (msooseth) [inline doc/src/for-developers.md:26]:
```suggestion
This value can be dynamically adjusted for a test group or a specific test.
```

**Review by msooseth — APPROVED** [review]
Sounds good! I'd consider it also interesting to talk about the general framework, how things are done inside the system, such as Expr.

We could also link to the paper:

https://link.springer.com/content/pdf/10.1007/978-3-031-65627-9_22.pdf?pdf=inline%20link

And the presentation:

https://github.com/msooseth/hevm-presentation

Or at least one of the 4 compiled PDFs there. Could be helpful?

**Comment by blishko** [comment]
Sure, I definitely want to add more stuff gradually. I'll try to add the links and explain `Expr`.
I am currently looking into profiling :)

**Comment by msooseth** [comment]
Profiling? Hmmmmmmmm I mean, yes, but also hm. Like, the MOST of the time will be spent in the SMT solver. My guess is that 100x to 1000x of the time. So I am not 100% sure making hevm faster would make such a big difference? Or maybe I'm wrong?

**Comment by msooseth** [comment]
Note that for real-world contract, the difference is likely more than 1000x, and potentially in the range of 1e6 or more, at least in my guesstimate.

**Comment by blishko** [comment]
I am sure you are right.

**Review by msooseth — APPROVED** [review]
lgtm

**Comment by blishko** [comment]
@msooseth, I have added links to the paper and to your presentations.
Documenting the data structures, such as `Expr` is left for future work.

File: doc/src/SUMMARY.md [hunk 2]
```diff
@@ -2,6 +2,7 @@
 [Quick Installation](./install.md)
 [When to Use](./when-to-use.md)
 [Limitations and Workarounds](./limitations-and-workarounds.md)
+[For Developers](./for-developers.md)
```
Comment (msooseth) [inline doc/src/SUMMARY.md:5]:
```suggestion
[For hevm Developers](./for-developers.md)
```

Comment (msooseth) [reply doc/src/SUMMARY.md:5]: It will also necessitate a name update to the file.

Comment (blishko) [reply doc/src/SUMMARY.md]: Should be fixed now.

**Pull request merged**